pkg/mutation/mutation_function.go

```diff
@@ -4,54 +4,105 @@ import (
 	"fmt"
 
 	"github.com/open-policy-agent/gatekeeper/pkg/mutation/path/parser"
+	path "github.com/open-policy-agent/gatekeeper/pkg/mutation/path/tester"
 	"github.com/open-policy-agent/gatekeeper/pkg/mutation/types"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 )
 
-func Mutate(mutator types.Mutator, obj *unstructured.Unstructured) error {
-	return mutate(mutator, obj.Object, nil, 0)
+func mutate(mutator types.Mutator, tester *path.Tester, obj *unstructured.Unstructured) error {
+	s := &mutatorState{mutator: mutator, tester: tester}
+	_, _, err := s.mutateInternal(obj.Object, 0)
+	return err
 }
 
-func mutate(m types.Mutator, current interface{}, previous interface{}, depth int) error {
-	if len(m.Path().Nodes)-1 == depth {
-		return addValue(m, current, previous, depth)
-	}
-	pathEntry := m.Path().Nodes[depth]
+type mutatorState struct {
+	mutator types.Mutator
+	tester  *path.Tester
+}
+
+// mutateInternal mutates the resource recursively. It returns false if there has been no change
+// to any downstream objects in the tree, indicating that the mutation should not be persisted
+func (s *mutatorState) mutateInternal(current interface{}, depth int) (bool, interface{}, error) {
+	pathEntry := s.mutator.Path().Nodes[depth]
```

Comment:
I think making sure len(path) != 0 is a good idea, as a mutation with no path makes no sense and we should catch that anywhere it leaks.

Looking at other validation code, it appears we are only validating whether `location` parses, which it would if it were empty. So, I think there are two work items:

1) Validate Mutate inputs using the minimum standards necessary for making a coherent mutate call
2) Add more `location` validation to Assign (I think AssignMeta enforces using a `metadata` prefix)

Reply:
SGTM. Assuming we apply additional validations earlier on, are you thinking we should still do bounds checking in places like this or build on the preconditions from validation?

Reply:
I think we should do "minimum sane input" bounds checking.

If there is no reason a function should see a given input (e.g. mutating a zero length path has no semantic meaning other than maybe overwriting the entire object), we should validate that so we can surface bugs like uninitialized structs.

I don't think we should go stricter than that though, to avoid needing to tweak validation in too many places if we add functionality.

Reply:
added the sanity checks